eva-accession-pipeline/src/main/java/uk/ac/ebi/eva/accession/pipeline/steps/tasklets/reportCheck/ReportCheckTasklet.java

```diff
@@ -0,0 +1,235 @@
+        vcfReader.open(new ExecutionContext());
+        reportReader.open(new ExecutionContext());
+        while (true) {
+            iterations++;
+
+            readOriginalVcfUntilBufferIsFull(variantBuffer);
+            removeMatchingVariants(variantBuffer, accessionBuffer);
+            boolean isEveryAccessionLoaded = readAccessionsAndRemoveMatchingOnes(variantBuffer, accessionBuffer);
+
+            maxAccessionBufferSize = Math.max(maxAccessionBufferSize, accessionBuffer.size());
+            if (isEveryAccessionLoaded) {
+                break;
+            }
+        }
+
+        logger.debug("Max unmatched accessions held in memory: {}; iterations: {}", maxAccessionBufferSize, iterations);
+        removeRemainingMatchingVariants(variantBuffer, accessionBuffer);
+
+        logStatus(stepContribution, variantBuffer, accessionBuffer);
+
+        return RepeatStatus.FINISHED;
+    }
+
+    /**
+     * Fill the variantBuffer (from the original VCF) until the maximum buffer size is
+     * reached, or until the file was completely read.
+     * @param variantBuffer buffer that will be filled with the original VCF
+     */
+    private void readOriginalVcfUntilBufferIsFull(Set<Variant> variantBuffer) throws Exception {
+        List<Variant> variantsRead;
+        while (variantBuffer.size() < maxVariantBufferSize && (variantsRead = readVcfIgnoringNonVariants()) != null) {
+            variantBuffer.addAll(variantsRead);
+        }
+    }
+
+    private List<Variant> readVcfIgnoringNonVariants() throws Exception {
+        List<Variant> variants = null;
+        boolean readVariantOrEof = false;
+        while (!readVariantOrEof) {
+            try {
+                variants = vcfReader.read();
+                readVariantOrEof = true;
+            } catch (Exception exception) {
+                if (skipPolicy.shouldSkip(exception, 0)) {
+                    // this was a non-variant, we must read the next line
+                } else {
+                    throw exception;
+                }
+            }
+        }
+        return variants;
+    }
+
+    private void removeMatchingVariants(Set<Variant> variantBuffer, Set<Variant> accessionBuffer) {
+        Set<Variant> matchedAccessions = new HashSet<>();
+        for (Variant unmatchedAccession : accessionBuffer) {
+            boolean removed = variantBuffer.remove(unmatchedAccession);
+            if (removed) {
+                matchedAccessions.add(unmatchedAccession);
+            }
+            if (variantBuffer.isEmpty()) {
+                break;
+            }
+        }
+        accessionBuffer.removeAll(matchedAccessions);
+    }
+
+    private boolean readAccessionsAndRemoveMatchingOnes(Set<Variant> variantBuffer, Set<Variant> accessionBuffer)
+            throws Exception {
+        Set<Variant> newUnmatchedAccessions = new HashSet<>();
+        List<Variant> reportVariantsRead = null;
+        newUnmatchedAccessions.clear();
```

Comment:
Were you trying to avoid creating a new set every time this method was called? Apparently, the performance of those two options is not so different and they recommend to use the one that allows a code easier to refactor. In this case, I would say that creating the set instead of reusing and clear one is cleaner.

Reply:
this version resulted after lots of refactorings. indeed, now it's useless to do a `.clear` just after a new instance is created. I'll remove the `clear`.